pkg/yurttunnel/util/util.go

```diff
@@ -52,41 +60,113 @@ func RunMetaServer(addr string) {
 	}()
 }
 
-// GetConfiguredDnatPorts returns the DNAT ports configured for tunnel server.
-// NOTE: We only allow user to add dnat rule that uses insecure port as the destination port currently.
-func GetConfiguredDnatPorts(client clientset.Interface, insecurePort string) ([]string, error) {
-	ports := make([]string, 0)
+// GetConfiguredProxyPortsAndMappings returns the proxy ports and mappings that configured for tunnel server.
+// field dnat-ports-pair will be deprecated in future version. it's recommended to use
+// field http-proxy-ports and https-proxy-ports.
+func GetConfiguredProxyPortsAndMappings(client clientset.Interface, insecureListenAddr, secureListenAddr string) ([]string, map[string]string, error) {
 	c, err := client.CoreV1().
 		ConfigMaps(YurttunnelServerDnatConfigMapNs).
 		Get(context.Background(), YurttunnelServerDnatConfigMapName, metav1.GetOptions{})
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			return nil, fmt.Errorf("configmap %s/%s is not found",
+			return []string{}, map[string]string{}, fmt.Errorf("configmap %s/%s is not found",
 				YurttunnelServerDnatConfigMapNs,
 				YurttunnelServerDnatConfigMapName)
 		} else {
-			return nil, fmt.Errorf("fail to get configmap %s/%s: %v",
+			return []string{}, map[string]string{}, fmt.Errorf("fail to get configmap %s/%s: %v",
 				YurttunnelServerDnatConfigMapNs,
 				YurttunnelServerDnatConfigMapName, err)
 		}
 	}
 
-	pairStr, ok := c.Data[yurttunnelServerDnatDataKey]
-	if !ok || len(pairStr) == 0 {
-		return ports, nil
+	return resolveProxyPortsAndMappings(c, insecureListenAddr, secureListenAddr)
+}
+
+// resolveProxyPortsAndMappings get proxy ports and port mappings from specified configmap
+func resolveProxyPortsAndMappings(cm *v1.ConfigMap, insecureListenAddr, secureListenAddr string) ([]string, map[string]string, error) {
+	portMappings := make(map[string]string, 0)
+	proxyPorts := make([]string, 0)
+
+	_, insecurePort, err := net.SplitHostPort(insecureListenAddr)
+	if err != nil {
+		return proxyPorts, portMappings, err
+	}
+
+	_, securePort, err := net.SplitHostPort(secureListenAddr)
+	if err != nil {
+		return proxyPorts, portMappings, err
+	}
+
+	// field dnat-ports-pair will be deprecated in future version
+	pairStr, ok := cm.Data[yurttunnelServerDnatDataKey]
```

Comment:
It seems that you can dedup the code 102-115 with 119-134 by introducing at local function.

Reply:
@Fei-Guo It's a good idea, fixed it.